Add script to automate release management

monfresh opened the pull request:
**Why**: Anything that is repeated manually more than a few times should
be automated. This also ensures everyone is performing release
management in the same way and reduces discrepancies and human error.

Hi! Before submitting your PR for review, and/or before merging it, please
go through the following checklist:

- [ ] For DB changes, check for missing indexes, check to see if the changes
affect other apps (such as the dashboard), make sure the DB columns in the
various environments are properly populated, coordinate with devops, plan
migrations in separate steps.

- [ ] For route changes, make sure GET requests don't change state or result in
destructive behavior. GET requests should only result in information being
read, not written.

- [ ] For encryption changes, make sure it is compatible with data that was
encrypted with the old code.

- [ ] For secrets changes, [make sure to update the S3 secrets bucket](https://github.com/18F/identity-private/wiki/Secrets-S3-buckets) with the 
new configs in **all** environments. 

- [ ] Do not disable Rubocop or Reek offenses unless you are absolutely sure
they are false positives. If you're not sure how to fix the offense, please
ask a teammate.

- [ ] When reading data, write tests for nil values, empty strings,
and invalid formats.

- [ ] When calling `redirect_to` in a controller, use `_url`, not `_path`.

- [ ] When adding user data to the session, use the `user_session` helper
instead of the `session` helper so the data does not persist beyond the user's
session.

- [ ] When adding a new controller that requires the user to be fully
authenticated, make sure to add `before_action :confirm_two_factor_authenticated`.

jmhooper commented on bin/release:
```diff
@@ -0,0 +1,80 @@
+  puts "==> This script will perform all of the following:"
+  puts "1. Checkout a new RC branch based on stages/int"
+  puts "2. Merge master into the RC branch"
+  puts "3. Force push the RC branch to stages/int on GitHub"
```
Why aren't we branching the RC off of master, and then merging the RC into stages/int? That would allow us to avoid a force push.

monfresh replied:
This was the result of the discussion we had around how to let people use the "squash and merge" feature. @tbaxter-18f did some research and concluded that in order to allow the "squash and merge" feature, and prevent merge conflicts on the `int`, `staging`, and `prod` branches when cherry picking commits, we had to branch the RC off of `int`, not `master`. The Release Management Wiki was updated a while back to reflect this, but I don't think everyone was following the new rules, which might explain why I had merge conflicts in `staging` and `prod` for RC 54 and other strange things like the diff being much larger than it should have been.

jmhooper commented on bin/release:
```diff
@@ -0,0 +1,116 @@
+end
+
+Dir.chdir ENV['HOME'] do
+  if `ls -d */`.include?('login-dot-gov')
```
This make some assumptions about how I have these repos saved on my machine. This will likely be pretty different for different developers who have their own preferences for how they organize things on their machine.

For example, all of my repos are in an `identity` folder and the names are the repo name on GH with the `identity-` lopped off, e.g. `identity/idp`, `identity/devops` etc. James has all of his in a `Code` directory with the original repo name as the filename.

monfresh replied:
The only assumption this is making is that you don't already have a directory in your home folder called `login-dot-gov`. It is creating a new directory called `login-dot-gov` within your home directory, which I thought would be an unlikely name that anyone is using. Inside this directory, it clones the devops and Wiki repos (without affecting the directories you use on a regular basis).

jmhooper replied:
Ah, I thought this was trying to update the repos I use for my work which would be frustrating.

Only thought from here is that is would be nice to use something like `/tmp/login-dot-gov` or `~/.tmp/login-dot-gov` instead, and also could it clean up after itself when it is done?

monfresh commented on bin/release:
```diff
@@ -0,0 +1,116 @@
+    puts "==> Cloning identity-devops repo into ~/login-dot-gov/identity-devops"
+    unless DRY_RUN == true
+      Dir.chdir "#{ENV['HOME']}/login-dot-gov" do
+        run "git clone https://github.com/18F/identity-devops.git"
```
Yes.

monfresh commented on bin/release:
```diff
@@ -0,0 +1,116 @@
+  rc_commit = `git log --grep="#{RC_BRANCH}" --oneline`
+  unless rc_commit.include?(RC_BRANCH)
+    puts "==> Generating the wiki page for the RC"
+    run "./scripts/generate-release --release-date=#{RELEASE_DATE}" unless DRY_RUN == true
```
Correct. That was more complicated to do as we would have to parse the page for the previous RC and insert the new link after it. This would also require that we know what the previous RC is, which we can't easily get because we are using different naming conventions for the RC branch and the number of the RC in the Wiki.

monfresh replied:
Or maybe there's a way to parse the Markdown and tell it to put it at the last line within the `Release candidates` bullet point.

jmhooper replied:
👍

monfresh commented:
@stevegsa I dropped the the `== true` on `unless DRY_RUN`. PTAL.
@jmhooper I switched to cloning via SSH, and I am now removing the `login-dot-gov` dir when the script is done.

I also added some new flags for deploying to the various envs. So, the sequence would be:

1 . Run `bin/release`. This will create the new RC branch, push it up to GitHub, and then open a PR against `stages/int`. Opening the PR is done via [hub](https://hub.github.com/). The first time you use hub, it will prompt you for your GitHub username, password, and 2FA code. After that, it will store a token in `~/.config/hub` and won't prompt you going forward.
2. Once you're ready to deploy to int, run `bin/release --int`. There are flags for staging and prod as well.

jmhooper commented on bin/release:
```diff
@@ -0,0 +1,160 @@
+end
+
+def open_pr_for_int
+  run "brew install hub" unless `brew list -1 | grep -Fqx hub`
```
FYI, this did not actually install hub, presumably because it exited with a zero status code?
